lib/World.luau

```diff
@@ -769,70 +865,173 @@ end
 ]=]
 
-function World:query(...)
-	debug.profilebegin("World:query")
-	assertValidComponent((...), 1)
+function World.query(self: World, ...)
+	-- TODO:
+	-- cache queries
+
+	--assertValidComponent((...), 1)
+	--print("Rebuilding query")
+	--print(self)
+	local components = { ... }
+	local a, b = ...
 
-	local metatables = { ... }
 	local queryLength = select("#", ...)
+	if queryLength == 1 then
+		components = { #a }
+	elseif queryLength == 2 then
+		components = { #a, #b }
+	else
+		error("Unimplemented query length")
+	end
 
-	local archetype = archetypeOf(...)
+	local possibleArchetypes
+	local compatibleArchetypes = {}
+	for _, componentId in components do
+		local associatedArchetypes = self.componentIndex[componentId]
+		if associatedArchetypes == nil then
+			error("No-op query unimplemented")
+		end
 
-	if self._queryCache[archetype] == nil then
-		self:_newQueryArchetype(archetype)
+		if possibleArchetypes == nil or #possibleArchetypes > #associatedArchetypes then
+			possibleArchetypes = associatedArchetypes
+		end
 	end
 
-	local compatibleArchetypes = self._queryCache[archetype]
+	-- Narrow the archetypes so only ones that contain all components are searched.
+	for archetypeId in possibleArchetypes do
+		local archetype = self.archetypes[archetypeId]
+		local incompatible = false
+		for _, componentId in components do
+			-- Does this archetype have this component?
+			if archetype.componentToStorageIndex[componentId] == nil then
+				-- Nope, so we can't use this one.
+				incompatible = true
+				break
+			end
+		end
 
-	debug.profileend()
+		if incompatible then
+			continue
+		end
 
-	if next(compatibleArchetypes) == nil then
-		-- If there are no compatible storages avoid creating our complicated iterator
-		return noopQuery
+		table.insert(compatibleArchetypes, archetype)
 	end
 
-	local queryOutput = table.create(queryLength)
+	local currentArchetypeIndex = 1
+	local currentArchetype = compatibleArchetypes[currentArchetypeIndex]
+	local currentArchetypeNumEntities = #currentArchetype.ownedEntities
+	local currentEntityIndex = 0
+	local function nextEntity()
+		currentEntityIndex += 1
+		if currentEntityIndex > currentArchetypeNumEntities then
+			currentEntityIndex = 0
+			currentArchetypeIndex += 1
+			currentArchetype = compatibleArchetypes[currentArchetypeIndex]
+			if currentArchetype == nil then
+				-- Out of entities.
+				return nil
+			end
 
-	local function expand(entityId, entityData)
-		if not entityId then
-			return
+			currentArchetypeNumEntities = #currentArchetype.ownedEntities
+			return nextEntity()
 		end
 
+		local entityId = currentArchetype.ownedEntities[currentEntityIndex]
+		local storage, componentToStorageIndex = currentArchetype.storage, currentArchetype.componentToStorageIndex
 		if queryLength == 1 then
-			return entityId, entityData[metatables[1]]
+			return entityId, storage[componentToStorageIndex[components[1]]][currentEntityIndex]
```

Comment:
This is way too expensive. Just reducing a single index would be a 20% performance increase. Jecs used to cache `componentStorageToIndex[components[N]]` but now that we are older and wiser.. You can just keep an upvalue for the column like jecs does which gave it a whooping 2x iteration speed increase.  

The code should just be `column[currentEntityIndex]`